Improve Google Cloud integration's error handling

noise-kngdm opened the pull request:
|Related issue|
|---|
|Closes #12140|

## Description
In this PR we rework the way the GCP module handles errors and exceptions. We add the `wodles/gcloud/exceptions.py` file, which enclose the information about all the errors the module might raise. All the exceptions are now being explicitely captured by the logger that is configured to set a different logging level depending on the severity of the error. Exceptions that are unhandled will be registered with the full exception trace to help debugging them. 

Along with this error handling system, we start handling errors that could be raised when specifying invalid credentials, bucket names and subscription and project names to make it easier for the user to use the module.

Because of the tests, we found out that the Access Logs integrations could be executed specifying '' as the bucket name, which made the module perform an incorrect API call to the GCP platform. We fixed that problem for this issue too by ensuring that the bucket name is of str type and that it evaluates to True, which means is not an empty string.

## Tests performed
All the tests passed without any further problems.

### Unit tests
We added unit tests to check that the changes that doesn't depend on the GCP API work as expected. All of them passed.

```
(wazuh-unit) ➜  gcloud git:(feature/12140-improve-gcp-error-handling) ✗ pytest
==================================== test session starts ====================================
platform linux -- Python 3.10.2, pytest-6.2.5, py-1.11.0, pluggy-0.13.1
rootdir: /home/gonzz/git/wazuh/wodles/gcloud
plugins: asyncio-0.15.1
collected 17 items                                                                          

tests/test_access_logs.py .                                                           [  5%]
tests/test_bucket.py ...                                                              [ 23%]
tests/test_gcloud.py .....                                                            [ 52%]
tests/test_integration.py .....                                                       [ 82%]
tests/test_subscriber.py ...                                                          [100%]

===================================== warnings summary ======================================
../../../../venv/wazuh-unit/lib/python3.10/site-packages/google/pubsub_v1/services/publisher/client.py:17
  /home/gonzz/venv/wazuh-unit/lib/python3.10/site-packages/google/pubsub_v1/services/publisher/client.py:17: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
    from distutils import util

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=============================== 17 passed, 1 warning in 0.04s ===============================
```

### Test Access Logs

<details><summary>Tests output</summary>

```
root@97cff1b2dfb7:/# /var/ossec/wodles/gcloud/gcloud -b framework-test-bucket  -c /var/ossec/wodles/gcloud/credentials.json -l 1 -o 2021-jan-01 -T access_logs
gcloud_wodle - INFO - Processing test_1_new.txt
gcloud_wodle - DEBUG - Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GOOGLE CLOUD STORAGE TEST": "Log 1", "source": "gcp_bucket"}}'"
gcloud_wodle - INFO - Processing test_22.txt
gcloud_wodle - INFO - Processing test_2_new.txt
gcloud_wodle - DEBUG - Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GOOGLE CLOUD STORAGE TEST": "Log 2", "source": "gcp_bucket"}}'"
gcloud_wodle - INFO - Processing test_3.txt
gcloud_wodle - DEBUG - Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GOOGLE CLOUD STORAGE TEST": "Log 3", "source": "gcp_bucket"}}'"
gcloud_wodle - INFO - Processing test_3_new.txt
gcloud_wodle - DEBUG - Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GOOGLE CLOUD STORAGE TEST": "Log 3", "source": "gcp_bucket"}}'"
gcloud_wodle - INFO - Processing test_4.txt
gcloud_wodle - DEBUG - Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GOOGLE CLOUD STORAGE TEST": "Log 4", "source": "gcp_bucket"}}'"
gcloud_wodle - INFO - Processing test_5.txt
gcloud_wodle - DEBUG - Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GOOGLE CLOUD STORAGE TEST": "Log 5", "source": "gcp_bucket"}}'"
gcloud_wodle - INFO - Processing test_6.txt
gcloud_wodle - DEBUG - Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GOOGLE CLOUD STORAGE TEST": "Log 6", "source": "gcp_bucket"}}'"
gcloud_wodle - INFO - Updating previously processed files.
gcloud_wodle - INFO - Received 7 messages
root@d3599e41fbc1:/# /var/ossec/wodles/gcloud/gcloud -b framework-test-bucket  -c /var/ossec/wodles/gcloud/credentials.json -l 1 -o 2021-jan-01 -T access_log 
gcloud_wodle - ERROR - An exception happened while running the wodle: Unsupported gcloud integration type: "access_log". The supported types are ('pubsub', 'access_logs')
Traceback (most recent call last):
  File "/var/ossec/wodles/gcloud/gcloud.py", line 89, in <module>
    raise exceptions.GCloudIntegrationTypeError(arguments.integration_type)
exceptions.GCloudIntegrationTypeError: Unsupported gcloud integration type: "access_log". The supported types are ('pubsub', 'access_logs')
root@d3599e41fbc1:/# /var/ossec/wodles/gcloud/gcloud -b framework-test-bucket  -c /var/ossec/wodles/gcloud/credentials.json -l 2 -o 2021-jan-01 -T access_log
gcloud_wodle - ERROR - An exception happened while running the wodle: Unsupported gcloud integration type: "access_log". The supported types are ('pubsub', 'access_logs')
root@d3599e41fbc1:/# /var/ossec/wodles/gcloud/gcloud -b framework-test-bucket  -c /var/ossec/wodles/gcloud/credentials.json -l 2 -o 2021-9jan-01 -T access_logs
usage: usage: gcloud.py [options]
gcloud.py: error: argument -o/--only_logs_after: Argument not a valid date in format YYYY-MMM-DD: '2021-9jan-01'.
root@d3599e41fbc1:/# /var/ossec/wodles/gcloud/gcloud -b framework-test-bucketo  -c /var/ossec/wodles/gcloud/credentials.json -l 2 -o 2021-jan-01 -T access_logs
gcloud_wodle - ERROR - An exception happened while running the wodle: The bucket 'framework-test-bucketo' does not exist
```

</details>



### Test Pub/Sub
<details><summary>Test output</summary>

```
root@97cff1b2dfb7:/# /var/ossec/wodles/gcloud/gcloud -T pubsub -l 1 -s test_gcloud_blog -p wazuh-dev-123456 -c /var/ossec/wodles/gcloud/credentials.json  
gcloud_wodle - DEBUG - Setting 1 thread to pull 100 messages in total
gcloud_wodle - INFO - Received and acknowledged 0 messages
root@d3599e41fbc1:/# /var/ossec/wodles/gcloud/gcloud -T pubsub -l 2 -s wazuh-audit-log -p wazuh-dev-123456 -c /var/ossec/wodles/gcloud/credentials.json -m 10
gcloud_wodle - ERROR - An exception happened while running the wodle: The 'wazuh-audit-log' project ID is incorrect or the user does not have permissions to access to it
root@d3599e41fbc1:/# /var/ossec/wodles/gcloud/gcloud -T pubsub -l 2 -s wazuh-audit-logs -p wazuh-dev-123456 -c /var/ossec/wodles/gcloud/credentials.json -m 10
gcloud_wodle - INFO - Received and acknowledged 10 messages
root@d3599e41fbc1:/# /var/ossec/wodles/gcloud/gcloud -T pubsub -l 2 -s wazuh-audit-logs -p wazuh-dev-123456 -c /var/ossec/wodles/gcloud/credentials.json -m 10 -t 99999
gcloud_wodle - WARNING - Reached maximum number of threads. Truncating to 80.
gcloud_wodle - INFO - Received and acknowledged 10 messages
root@d3599e41fbc1:/# /var/ossec/wodles/gcloud/gcloud -T pubsub -l 2 -s wazuh-audit-logs -p wazuh-dev-123456 -c /var/ossec/wodles/gcloud/credentials.json -m 10 -t -99999
gcloud_wodle - ERROR - An exception happened while running the wodle: The minimum number of threads is 1. Please check your configuration.
root@d3599e41fbc1:/# /var/ossec/wodles/gcloud/gcloud -T pubsub -l 2 -s wazuh-audit-logs -p wazuh-dev-1234568 -c /var/ossec/wodles/gcloud/credentials.json -m 10     
gcloud_wodle - ERROR - An exception happened while running the wodle: The 'wazuh-dev-1234568' subscription is incorrect or the user credentials are not valid
root@d3599e41fbc1:/# /var/ossec/wodles/gcloud/gcloud -T pubsub -l 1 -s wazuh-audit-logs -p wazuh-dev-1234568 -c /var/ossec/wodles/gcloud/credentials.json -m 10 
gcloud_wodle - DEBUG - Setting 1 thread to pull 10 messages in total
gcloud_wodle - ERROR - An exception happened while running the wodle: The 'wazuh-dev-1234568' subscription is incorrect or the user credentials are not valid
Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/site-packages/google/api_core/grpc_helpers.py", line 67, in error_remapped_callable
    return callable_(*args, **kwargs)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/grpc/_channel.py", line 946, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/grpc/_channel.py", line 849, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.NOT_FOUND
	details = "Requested project not found or user does not have access to it (project=wazuh-dev-1234568). Make sure to specify the unique project identifier and not the Google Cloud Console display name."
	debug_error_string = "{"created":"@1646826490.039509764","description":"Error received from peer ipv4:142.250.184.10:443","file":"src/core/lib/surface/call.cc","file_line":1067,"grpc_message":"Requested project not found or user does not have access to it (project=wazuh-dev-1234568). Make sure to specify the unique project identifier and not the Google Cloud Console display name.","grpc_status":5}"
>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/var/ossec/wodles/gcloud/pubsub/subscriber.py", line 91, in check_permissions
    response = self.subscriber.test_iam_permissions(
  File "/var/ossec/framework/python/lib/python3.9/site-packages/google/cloud/pubsub_v1/_gapic.py", line 40, in <lambda>
    fx = lambda self, *a, **kw: wrapped_fx(self.api, *a, **kw)  # noqa
  File "/var/ossec/framework/python/lib/python3.9/site-packages/google/pubsub_v1/services/subscriber/client.py", line 2029, in test_iam_permissions
    response = rpc(request, retry=retry, timeout=timeout, metadata=metadata,)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/google/api_core/gapic_v1/method.py", line 145, in __call__
    return wrapped_func(*args, **kwargs)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/google/api_core/grpc_helpers.py", line 69, in error_remapped_callable
    six.raise_from(exceptions.from_grpc_error(exc), exc)
  File "<string>", line 3, in raise_from
google.api_core.exceptions.NotFound: 404 Requested project not found or user does not have access to it (project=wazuh-dev-1234568). Make sure to specify the unique project identifier and not the Google Cloud Console display name.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/ossec/wodles/gcloud/gcloud.py", line 62, in <module>
    subscriber_client.check_permissions()
  File "/var/ossec/wodles/gcloud/pubsub/subscriber.py", line 97, in check_permissions
    raise exceptions.GCloudPubSubSubscriptionError(
exceptions.GCloudPubSubSubscriptionError: The 'wazuh-dev-1234568' subscription is incorrect or the user credentials are not valid
```

</details>

CarlosRS9 reviewed (changes requested):
I think we should use an approach similar to the one used in exception handling in the Framework. Instead of defining so many separate classes in `exceptions.py` we should have only 2 or 3 and use a dictionary with the relevant information, simplifying the design of the classes.

noise-kngdm commented:
## Update
I modified the design of the exceptions as requested. Since the changes affected almost everything I had done, I performed the tests again.

## Tests performed
All the tests passed without any further problems.

### Unit tests
We added unit tests to check that the changes that doesn't depend on the GCP API work as expected. All of them passed.

```
(wazuh-unit) ➜  gcloud git:(feature/12140-improve-gcp-error-handling) ✗ pytest
==================================== test session starts ====================================
platform linux -- Python 3.10.2, pytest-6.2.5, py-1.11.0, pluggy-0.13.1
rootdir: /home/gonzz/git/wazuh/wodles/gcloud
plugins: asyncio-0.15.1
collected 17 items                                                                          

tests/test_access_logs.py .                                                           [  5%]
tests/test_bucket.py ...                                                              [ 23%]
tests/test_gcloud.py .....                                                            [ 52%]
tests/test_integration.py .....                                                       [ 82%]
tests/test_subscriber.py ...                                                          [100%]

===================================== warnings summary ======================================
../../../../venv/wazuh-unit/lib/python3.10/site-packages/google/pubsub_v1/services/publisher/client.py:17
  /home/gonzz/venv/wazuh-unit/lib/python3.10/site-packages/google/pubsub_v1/services/publisher/client.py:17: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
    from distutils import util

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=============================== 17 passed, 1 warning in 0.04s ===============================
```

### Test Access Logs

<details><summary>Tests output</summary>

```
root@97cff1b2dfb7:/# /var/ossec/wodles/gcloud/gcloud -b framework-test-bucket  -c /var/ossec/wodles/gcloud/credentials.json -l 1 -o 2021-jan-01 -T access_logs
gcloud_wodle - INFO - Processing test_1_new.txt
gcloud_wodle - DEBUG - Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GOOGLE CLOUD STORAGE TEST": "Log 1", "source": "gcp_bucket"}}'"
gcloud_wodle - INFO - Processing test_22.txt
gcloud_wodle - INFO - Processing test_2_new.txt
gcloud_wodle - DEBUG - Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GOOGLE CLOUD STORAGE TEST": "Log 2", "source": "gcp_bucket"}}'"
gcloud_wodle - INFO - Processing test_3.txt
gcloud_wodle - DEBUG - Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GOOGLE CLOUD STORAGE TEST": "Log 3", "source": "gcp_bucket"}}'"
gcloud_wodle - INFO - Processing test_3_new.txt
gcloud_wodle - DEBUG - Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GOOGLE CLOUD STORAGE TEST": "Log 3", "source": "gcp_bucket"}}'"
gcloud_wodle - INFO - Processing test_4.txt
gcloud_wodle - DEBUG - Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GOOGLE CLOUD STORAGE TEST": "Log 4", "source": "gcp_bucket"}}'"
gcloud_wodle - INFO - Processing test_5.txt
gcloud_wodle - DEBUG - Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GOOGLE CLOUD STORAGE TEST": "Log 5", "source": "gcp_bucket"}}'"
gcloud_wodle - INFO - Processing test_6.txt
gcloud_wodle - DEBUG - Sending msg to analysisd: "b'1:Wazuh-GCloud:{"integration": "gcp", "gcp": {"GOOGLE CLOUD STORAGE TEST": "Log 6", "source": "gcp_bucket"}}'"
gcloud_wodle - INFO - Updating previously processed files.
gcloud_wodle - INFO - Received 7 messages

root@d3599e41fbc1:/# /var/ossec/wodles/gcloud/gcloud -b framework-test-bucket  -c /var/ossec/wodles/gcloud/credentials.json -l 1 -o 2021-jan-01 -T access_log 
gcloud_wodle - ERROR - An exception happened while running the wodle: GCloudIntegrationTypeError: Unsupported gcloud integration type: "access_log".The supported types are ('pubsub', 'access_logs')
Traceback (most recent call last):
  File "/var/ossec/wodles/gcloud/gcloud.py", line 92, in main
    raise exceptions.GCloudError(3, integration_type=arguments.integration_type)
exceptions.GCloudError: GCloudIntegrationTypeError: Unsupported gcloud integration type: "access_log".The supported types are ('pubsub', 'access_logs')

root@4926b5380b7e:/var/ossec/wodles/gcloud# /var/ossec/wodles/gcloud/gcloud -b framework-test-bucket  -c /var/ossec/wodles/gcloud/credentials.json -l 3 -o 2021-jan-01 -T access_log 
gcloud_wodle - ERROR - An exception happened while running the wodle: GCloudIntegrationTypeError: Unsupported gcloud integration type: "access_log".The supported types are ('pubsub', 'access_logs')

root@d3599e41fbc1:/# /var/ossec/wodles/gcloud/gcloud -b framework-test-bucket  -c /var/ossec/wodles/gcloud/credentials.json -l 2 -o 2021-9jan-01 -T access_logs
usage: usage: gcloud.py [options]
gcloud.py: error: argument -o/--only_logs_after: Argument not a valid date in format YYYY-MMM-DD: '2021-9jan-01'.

root@d3599e41fbc1:/# /var/ossec/wodles/gcloud/gcloud -b framework-test-bucketo  -c /var/ossec/wodles/gcloud/credentials.json -l 2 -o 2021-jan-01 -T access_logs
gcloud_wodle - ERROR - An exception happened while running the wodle: GCloudBucketNotFound: The bucket 'framework-test-bucketo' does not exist
```

</details>



### Test Pub/Sub
<details><summary>Test output</summary>

```
root@97cff1b2dfb7:/# /var/ossec/wodles/gcloud/gcloud -T pubsub -l 1 -s test_gcloud_blog -p wazuh-dev-123456 -c /var/ossec/wodles/gcloud/credentials.json  
gcloud_wodle - DEBUG - Setting 1 thread to pull 100 messages in total
...
gcloud_wodle - INFO - Received and acknowledged 100 messages

root@d3599e41fbc1:/# /var/ossec/wodles/gcloud/gcloud -T pubsub -l 2 -s wazuh-audit-log -p wazuh-dev-123456 -c /var/ossec/wodles/gcloud/credentials.json -m 10
gcloud_wodle - ERROR - An exception happened while running the wodle: GCloudPubSubProjectError: The 'wazuh-audit-log' project ID is incorrect or the user does not have permissions to access to it

root@d3599e41fbc1:/# /var/ossec/wodles/gcloud/gcloud -T pubsub -l 2 -s wazuh-audit-logs -p wazuh-dev-123456 -c /var/ossec/wodles/gcloud/credentials.json -m 10
gcloud_wodle - INFO - Received and acknowledged 10 messages

root@d3599e41fbc1:/# /var/ossec/wodles/gcloud/gcloud -T pubsub -l 2 -s wazuh-audit-logs -p wazuh-dev-123456 -c /var/ossec/wodles/gcloud/credentials.json -m 10 -t 99999
gcloud_wodle - WARNING - Reached maximum number of threads. Truncating to 80.
gcloud_wodle - INFO - Received and acknowledged 10 messages

root@4926b5380b7e:/var/ossec/wodles/gcloud# /var/ossec/wodles/gcloud/gcloud -T pubsub -l 2 -s wazuh-audit-logs -p wazuh-dev-123456 -c /var/ossec/wodles/gcloud/credentials.json -m 10 -t -99999
gcloud_wodle - ERROR - An exception happened while running the wodle: GCloudPubSubNumThreadsError: The minimum number of threads is 1. Please check your configuration.

root@4926b5380b7e:/var/ossec/wodles/gcloud# /var/ossec/wodles/gcloud/gcloud -T pubsub -l 2 -s wazuh-audit-logs -p wazuh-dev-1234568 -c /var/ossec/wodles/gcloud/credentials.json -m 10
gcloud_wodle - ERROR - An exception happened while running the wodle: GCloudPubSubSubscriptionError: The 'wazuh-dev-1234568' subscription is incorrect or the user credentials are not valid

root@d3599e41fbc1:/# /var/ossec/wodles/gcloud/gcloud -T pubsub -l 1 -s wazuh-audit-logs -p wazuh-dev-1234568 -c /var/ossec/wodles/gcloud/credentials.json -m 10 
gcloud_wodle - DEBUG - Setting 1 thread to pull 10 messages in total
gcloud_wodle - ERROR - An exception happened while running the wodle: GCloudPubSubSubscriptionError: The 'wazuh-dev-1234568' subscription is incorrect or the user credentials are not valid
Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/site-packages/google/api_core/grpc_helpers.py", line 67, in error_remapped_callable
    return callable_(*args, **kwargs)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/grpc/_channel.py", line 946, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/grpc/_channel.py", line 849, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.NOT_FOUND
	details = "Requested project not found or user does not have access to it (project=wazuh-dev-1234568). Make sure to specify the unique project identifier and not the Google Cloud Console display name."
	debug_error_string = "{"created":"@1647272658.083579005","description":"Error received from peer ipv4:172.217.17.10:443","file":"src/core/lib/surface/call.cc","file_line":1067,"grpc_message":"Requested project not found or user does not have access to it (project=wazuh-dev-1234568). Make sure to specify the unique project identifier and not the Google Cloud Console display name.","grpc_status":5}"
>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/var/ossec/wodles/gcloud/pubsub/subscriber.py", line 102, in check_permissions
    response = self.subscriber.test_iam_permissions(
  File "/var/ossec/framework/python/lib/python3.9/site-packages/google/cloud/pubsub_v1/_gapic.py", line 40, in <lambda>
    fx = lambda self, *a, **kw: wrapped_fx(self.api, *a, **kw)  # noqa
  File "/var/ossec/framework/python/lib/python3.9/site-packages/google/pubsub_v1/services/subscriber/client.py", line 2029, in test_iam_permissions
    response = rpc(request, retry=retry, timeout=timeout, metadata=metadata,)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/google/api_core/gapic_v1/method.py", line 145, in __call__
    return wrapped_func(*args, **kwargs)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/google/api_core/grpc_helpers.py", line 69, in error_remapped_callable
    six.raise_from(exceptions.from_grpc_error(exc), exc)
  File "<string>", line 3, in raise_from
google.api_core.exceptions.NotFound: 404 Requested project not found or user does not have access to it (project=wazuh-dev-1234568). Make sure to specify the unique project identifier and not the Google Cloud Console display name.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/ossec/wodles/gcloud/gcloud.py", line 65, in main
    subscriber_client.check_permissions()
  File "/var/ossec/wodles/gcloud/pubsub/subscriber.py", line 108, in check_permissions
    raise exceptions.GCloudError(204, subscription=self.subscription_path.split('/')[1])
exceptions.GCloudError: GCloudPubSubSubscriptionError: The 'wazuh-dev-1234568' subscription is incorrect or the user credentials are not valid
```

</details>

CarlosRS9 commented on wodles/gcloud/tests/test_integration.py:
```diff
@@ -54,3 +56,34 @@ def test_format_msg():
+
+
+@pytest.mark.parametrize('raised_exception, expected_exception, errcode', [
+    (ConnectionRefusedError, exceptions.GCloudInternalError, 800),
+    (OSError, exceptions.GCloudInternalError, 801)
```
```suggestion
    (ConnectionRefusedError, exceptions.WazuhIntegrationInternalError, 1),
    (OSError, exceptions.WazuhIntegrationInternalError, 2)
```

CarlosRS9 commented on wodles/gcloud/tests/test_integration.py:
```diff
@@ -54,3 +56,34 @@ def test_format_msg():
+])
+def test_initialize_socket_ko(
+        raised_exception: Exception,
+        expected_exception: exceptions.GCloudException,
```
```suggestion
        expected_exception: exceptions.WazuhIntegrationException,
```

CarlosRS9 commented on wodles/gcloud/tests/test_integration.py:
```diff
@@ -54,3 +56,34 @@ def test_format_msg():
+    ----------
+    raised_exception : Exception
+        Exception raised that will be captured.
+    expected_exception : exceptions.GCloudException
```
```suggestion
    expected_exception : exceptions.WazuhIntegrationException
```

CarlosRS9 commented on wodles/gcloud/tests/test_subscriber.py:
```diff
@@ -10,34 +10,64 @@
+    ])
+def test_subscription_ko(credentials_file: str, logger: Logger,
+                         project: str, subscription_id: str,
+                         exception: exceptions.GCloudException,
```
```suggestion
                         exception: exceptions.WazuhIntegrationException,
```

CarlosRS9 commented on wodles/gcloud/tests/test_subscriber.py:
```diff
@@ -10,34 +10,64 @@
+        Name of the project.
+    subscription_id : str
+        ID of the subscription.
+    exception : exceptions.GCloudException
```
```suggestion
    exception : exceptions.WazuhIntegrationException
```

CarlosRS9 commented on wodles/gcloud/buckets/bucket.py:
```diff
@@ -47,11 +48,22 @@ def __init__(self, credentials_file: str, logger: logging.Logger, bucket_name: s
+
+        Raises
+        ------
+        GCloudError
```
```suggestion
        exceptions.GCloudError
```

CarlosRS9 commented on wodles/gcloud/buckets/bucket.py:
```diff
@@ -182,14 +193,22 @@ def _get_last_creation_time(self):
+
+        Raises
+        ------
+        GCloudError
```
```suggestion
        exceptions.GCloudError
```

CarlosRS9 reviewed (approved):
LGTM!
